File: ydb/benches/sdk_compare/result.rs
use serde::Serialize;

use crate::config::Scenario;
use crate::metrics::LatencyMetric;

#[derive(Debug, Serialize)]
pub(crate) struct BenchmarkResult {
    scenario: Scenario,
    implementation: Implementation,
    metrics: BenchmarkMetrics,
}

impl BenchmarkResult {
    pub(crate) fn topic(scenario: Scenario, metrics: TopicMetrics) -> Self {
        Self {
            scenario,
            implementation: Implementation::rust(),
            metrics: BenchmarkMetrics::Topic(metrics),
        }
    }

    pub(crate) fn query(scenario: Scenario, metrics: QueryMetrics) -> Self {
        Self {
            scenario,
            implementation: Implementation::rust(),
            metrics: BenchmarkMetrics::Query(metrics),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum BenchmarkMetrics {
    Topic(TopicMetrics),
    Query(QueryMetrics),
}

#[derive(Debug, Serialize)]
struct Implementation {
    language: &'static str,
    sdk_version: &'static str,
    build_profile: &'static str,
}

impl Implementation {
    fn rust() -> Self {
        Self {
            language: "rust",
            sdk_version: env!("CARGO_PKG_VERSION"),
            build_profile: if cfg!(debug_assertions) {
                "debug"
            } else {
                "release"
            },
        }
    }
}

#[derive(Debug, Serialize)]
pub(crate) struct TopicMetrics {
    #[serde(rename = "topic.write_ack")]
    pub(crate) write_ack: LatencyMetric,
    #[serde(rename = "topic.end_to_end")]
    pub(crate) end_to_end: LatencyMetric,
    #[serde(rename = "topic.commit_ack")]
    pub(crate) commit_ack: LatencyMetric,
    pub(crate) write_messages_per_second: f64,
    pub(crate) write_bytes_per_second: f64,
    pub(crate) read_messages_per_second: f64,
    pub(crate) read_bytes_per_second: f64,
}

#[derive(Debug, Serialize)]
pub(crate) struct QueryMetrics {
    #[serde(rename = "query.execute")]
    pub(crate) execute: LatencyMetric,
    pub(crate) queries_per_second: f64,
    pub(crate) rows_per_second: f64,
    pub(crate) payload_bytes_per_second: f64,
}

File: benchmarks/sdk-compare/cpp/src/core/result.cpp
#include "result.h"

#include <ydb-cpp-sdk/client/resources/ydb_resources.h>

namespace sdk_compare {

void to_json(nlohmann::json& value, const TopicMetrics& metrics) {
    value = nlohmann::json{
        {"topic.write_ack", metrics.write_ack},
        {"topic.end_to_end", metrics.end_to_end},
        {"topic.commit_ack", metrics.commit_ack},
        {"write_messages_per_second", metrics.write_messages_per_second},
        {"write_bytes_per_second", metrics.write_bytes_per_second},
        {"read_messages_per_second", metrics.read_messages_per_second},
        {"read_bytes_per_second", metrics.read_bytes_per_second},
    };
}

void to_json(nlohmann::json& value, const BenchmarkResult& result) {
    value = nlohmann::json{
        {"scenario", result.scenario},
        {"implementation", result.implementation},
        {"metrics", result.metrics},
    };
}

Implementation cpp_implementation() {
    return {
        .language = "cpp",
        .sdk_version = NYdb::GetSdkSemver(),
#ifdef NDEBUG
        .build_profile = "release",
#else
        .build_profile = "debug",
#endif
    };
}

}  // namespace sdk_compare
